enhancements/compact-clusters.md

```diff
@@ -0,0 +1,286 @@
+
+The primary limitation blocking this in 4.2 and earlier was a limitation in
+Kubernetes that prevented service load balancers from targeting the master nodes
+when the node-role for masters was set. This was identified as a bug and in
+Kube 1.16 we [clarified that this should be fixed by moving the functionality to
+explicit labels](https://github.com/kubernetes/enhancements/issues/1143) and
+[added two new alpha labels](https://github.com/kubernetes/kubernetes/pull/80238)
+that made the old behavior still achievable. The new alpha feature gate `LegacyNodeRoleBehavior`
+disables the old check, which would allow service load balancers to target masters.
+The risk of regression is low because the code path is simple, and all usage of
+the gate would be internal. We would enable this gate by default in 4.3.
+
+The IPI and UPI install and documentation must ensure that service load balancers
+can correctly target master machines if a 3-node cluster is created by ensuring
+security groups are correctly targeted. It may be desirable to switch the default
+ingress port policy to be NodePort to remove the need to listen on the host in
+these environments in 4.3 across all appropriate clouds, but that may not be
+strictly required.
+
+In order to ensure this is successful and testable by default, some of the default
+e2e-aws runs should be switched to a 3-node cluster configuration, no worker pool
+defined. Tests that fail should be corrected. Resource limits that are insufficient
+to handle the workload should be bumped. Any hardcoded logic that breaks in this
+configuration should be fixed.
+
+At the end of this work, we should be able to declare a 3-node 4.3 cluster in the
+cloud as fully supported.
+
+### Support for one-node control planes
+
+OpenShift 3.x supported single node "all-in-one" clusters via both `oc cluster up`,
+`openshift-ansible`, and other more opinionated configurations like pre-baked VMs.
+OpenShift 4, with its different control plane strategy, intentionally diverges from
+the static preconfiguration of 3.x but is geared towards highly available control
+planes and upgrades. To tolerate single control-plane node upgrades we would need
+to fix a number of assumptions in core operators. In addition, we wish to introduce
+the cluster-etcd-operator to allow for master replacement, and the capabilities
+that operator will introduce 
+
+For the 4.3 timeframe the following statements are true for single-node clusters:
+
+* The installer allows creation of single-master control planes
+* Some components may not completely roll out because they require separate nodes
+  * Users may be required to disable certain operators and override their replicas count
+* Upgrades may hang or pause because operators assume at least 2 control plane members are available at all times
+
+In general it should be possible with some work to get a single-node cluster but
+it is not support for upgrade or production use.
```

Comment:
Let's try to come up with a different term than "single node cluster" because it's a contradiction.  How about "single node" or "single non-clustered"?

Also...I don't quite agree with people who are saying single node *can't* be upgraded.  This is software we're talking about, most things are *possible* - it's a question of tradeoffs.

We know for sure that single node can't be upgraded while maintaining availability - that's obvious.   And we know that the MCO at least would need work in this scenario - it'd need to learn to skip draining for example.  But I wonder whether it'd really require more than the MCO - I mean, we're skipping all of the other operator's PDBs etc.

Reply:
In other words, someone tell me why this is actually any harder than having the MCO skip draining?  I may just try patching the code to do that and see...

Reply:
> Also...I don't quite agree with people who are saying single node can't be upgraded.

i think those people mean "can't be upgraded as the product exists today" (ie w/o code changes).

Reply:
BTW on the single node topic...worth linking to https://github.com/rancher/k3s as I think they're doing interesting stuff there.  Replacing etcd with sqlite does kind of just make sense in this scenario etc.

Reply:
No, it really doesn’t.  K3s has a number of serious challenges and is essentially “not kube”.  You can’t replace etcd and pretend it’s still kube without obeying all semantics, and the choices are questionable.

Single node clusters that are custom built to target single machines aren’t really of relevance here - the benefits in resource usage come with significant costs and downsides that don’t really apply to the core value prop of OpenShift.

Reply:
This statement is that right now you can’t upgrade OpenShift on a single node.  It has nothing to do with whether it is possible.

Reply:
> K3s has a number of serious challenges and is essentially “not kube”. 

I think that's part of the idea behind the name.  And we have a similar situation with CodeReady Containers - it's not "branded" OpenShift (not directly).

Maybe we should think of these non-cluster cases as more "reusing OpenShift technology". To say it another way: this enhancement is not about making compact OpenShift - it's about helping other projects use our components for non-cluster use cases.  Something like that?

Reply:
This enhancement is about compact openshift.  We're just rejecting single node openshift as a short term target because it doesn't satisfy the goals we have to for resiliency, reaction to change, etc.